src/Models/UserIdentityModel.php

```diff
@@ -211,6 +214,140 @@ public function getAllAccessTokens(User $user): array
             ->findAll();
     }
 
+    // HMAC
+    /**
+     * Find and Retrieve the HMAC AccessToken based on Token alone
+     *
+     * @return ?AccessToken Full HMAC Access Token object
+     */
+    public function getHmacTokenByKey(string $key): ?AccessToken
+    {
+        return $this
+            ->where('type', HmacSha256::ID_TYPE_HMAC_TOKEN)
+            ->where('secret', $key)
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * Generates a new personal access token for the user.
+     *
+     * @param string   $name   Token name
+     * @param string[] $scopes Permissions the token grants
+     *
+     * @throws Exception
+     * @throws ReflectionException
+     */
+    public function generateHmacToken(User $user, string $name, array $scopes = ['*']): AccessToken
+    {
+        $this->checkUserId($user);
+
+        $return = $this->insert([
+            'type'    => HmacSha256::ID_TYPE_HMAC_TOKEN,
+            'user_id' => $user->id,
+            'name'    => $name,
+            'secret'  => bin2hex(random_bytes(16)), // Key
+            'secret2' => bin2hex(random_bytes(16)), // Secret Key
```

Comment:
Why do you use 16 byte random data?
32 is better?

Reply:
Because `bin2hex(random_bytes(16))` produces a 32 character string.  The HMAC hash, combined with the 32 char string produces a lengthy header as is.  This is consistent with what I have seen used elsewhere, but if you feel a 32 byte binary (producing a 64 char string) will not over-complicate things or cause any issues with header length I can change it.   Perhaps the length should be configurable?

Reply:
I'm not a security expert, so may be wrong.
But in my understanding, the secret key length should have 32 bytes entropy.
Ref, https://crypto.stackexchange.com/questions/31473/what-size-should-the-hmac-key-be-with-sha-256

and even if we use 32 bytes for the secret key (secret2), the header won't be longer.

Reply:
@kenjis I believe you are correct here.  Let's keep the key at 32 char (16 bytes) and bump up the secret key byte size.  I am inclined to allow this to be configurable as I could see someone wanting to override the default.  I will add that to the new config file.